OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+
+    forceResAndDerivative =
+            calcFiberStaticEquilibResidual(xdx, muscleLength, activ);
+    bool largeDeriv=abs(forceResAndDerivative[0])> (0.5*abs(fx));
+
+    if (~inInterval || largeDeriv) {
+        x=(a+b)/2;
+        dx=(a-b)/2;}
+    else
+        {x=xdx;};
+
+
+    };
+
+    // TODO:  Need to handle condition when number of loop iterations reaches
+    // neval limit. See Millard2012 Muscle Error Handling
+
+    SimTK::Vec2 vout;
+    vout[0]=x;   //projFiberLengthNorm
+    vout[1]=forceResAndDerivative[2];  // muscleForce
+    return vout;
+}
+
+
+
+
+//------------------------------------------------------------------------------
+//Calculate the ydot values to drive the residuals to 0 and "balance" the muscle
+SimTK::Vector VandenBogert2011Muscle::calcSolveMuscle(SimTK::State& s,
+         double activ, SimTK::Vector yDotInitialGuess) const {
+
+
+    ImplicitSystemDerivativeSolver sys(2, s, activ);
+
+    //TODO:  Need come up with reasonable bounds
+    SimTK::Vector lower_bounds(2);
+    lower_bounds[0] = -SimTK::Infinity;
+    lower_bounds[1] = -SimTK::Infinity;
+
+    SimTK::Vector upper_bounds(2);
+    upper_bounds[0] = SimTK::Infinity;
+    upper_bounds[1] = SimTK::Infinity;
+
+    sys.setParameterLimits(lower_bounds, upper_bounds);
+
+    SimTK::Optimizer opt(sys, SimTK::InteriorPoint); //Create the optimizer
+
+    // Specify settings for the optimizer
+    opt.setConvergenceTolerance(0.1);
+    opt.useNumericalGradient(true, 1e-5);
+    opt.useNumericalJacobian(true);
+    opt.setMaxIterations(100);
+    opt.setLimitedMemoryHistory(500);
+
+    opt.optimize(yDotInitialGuess);  // Optimize
+
+    return yDotInitialGuess;
+}
+
+//------------------------------------------------------------------------------
+//Create a optimization solver to balance the muscle
+class ImplicitSystemDerivativeSolver : public SimTK::OptimizerSystem {
+public:
+
+    //TODO:  Add back in option to use NR solver (and perform benchmark)
+    /* Constructor class. Parameters passed are accessed in the objectiveFunc()
+     * class. */
+    ImplicitSystemDerivativeSolver(int numParameters, SimTK::State& s,
+                                   double& u):
+            OptimizerSystem(numParameters),
+            numControls(numParameters),
+            si(s),
+            excitation(u)
+    {}
+
+    int objectiveFunc(const SimTK::Vector &new_yDotGuess, bool new_params,
+                      SimTK::Real& f) const override {
+        // No objective function
+        f = 0;
+        return 0;
+    }
+
+    int constraintFunc(const SimTK::Vector &new_yDotGuess, bool newParams,
+                       SimTK::Vector& constraints)  override {
+        //The constraints function is the residuals
+
+        double projFibVelNorm_guess=new_yDotGuess[0];
+        double activdot_guess=new_yDotGuess[1];
+        VandenBogert2011Muscle::ImplicitResidual
+                Results = VandenBogert2011Muscle::calcImplicitResidual
```

Comment:
My guess is that this doesn't compile. This class will need a member variable `const VandenBogert2011Muscle&`, which can be set from the constructor.